### PR TITLE
Add pva links to pvxs

### PR DESCRIPTION
**Pull request by simon-ess**
Port PVA links from pva2pva.

TODO

- [X] Add `atomic` as link option.
- [X] Switch logger names to `pvxs.*`.
- [X] Test coverage for AfterPut
- [X] Test coverage for disconnected link
- [X] Fully resolve https://github.com/epics-base/epics-base/issues/423
- [X] Interoperability with QSRV1.  Automatically disable QSRV2 during iocInit when QSRV1 also present.
- [X] Clean up `pvxs/iochooks.h`.

### CODE REVIEW

**Comment by mdavidsaver** [comment]
I have pushed some additional changes.  Including a possible workaround for https://github.com/epics-base/epics-base/issues/423 with older Base (currently all released versions...).  I have also reworked the test synchronization code.  However, I am still seeing some intermittent test failures in GHA runs.

**Comment by mdavidsaver** [comment]
Another update.  At this point all tests are passing for me.  Although, coverage.sh shows there are a couple of cases which I think need testing.

wrt. the `db_cancel_event()` issue, I think I understand now what was happening.  The trigger was a self-cancel, which should not be possible.  However, due to an issue with the lifetime, objects bound into lambda functions were no released immediately when the associated subscription was canceled.  This should be fixed by 9b099be0d8d548da8ad6072529f5486426c0eac4.  I think this will act as a workaround for the Base issue.

The other lingering issue was an intermittent failure in `testSevr()` of testpvalink.  I think I traced this to an issue with the `pvaGlobal_t::channels` cache.  A cache hit which was processed before an additional data update did not call `pvaLink::onTypeChange()`, so it was using previous Value s.

**Comment by mdavidsaver** [comment]
@simon-ess I think I have this PR is a more or less final form.  At this point, the blocker to merging is a resolution of https://github.com/epics-base/epics-base/issues/423, which causes `testpvalink` to fail.

As a remaining task, I am working on some logic to detect when `qsrv.dbd` from pva2pva has also been included, and will use this to automatically disable QSRV2 during startup.  Of course, then absent QSRV2 would then default to automatically starting in full.  (like https://github.com/mdavidsaver/pvxs/pull/58)

**Review by simon-ess — COMMENTED** [review]
Also, one other oddity I noticed: if you have the following as a db file:
```
record(ai, "FOO") {
    field(INP, {pva: {pv: "BAR", proc: "CPP"}})
}

record(aai, "BAR") {
    field(NELM, 10)
}
```
then running `pvput BAR "[1,2,3]"` with an IOC running softIocPVA the linked `ai` record fetches the first value. However, with `softIocPVX` it does not.

**Comment by george-mcintyre** [comment]
Comments interspersed below.


> On 6 Nov 2023, at 17:09, Simon Rose ***@***.***> wrote:
> 
> 
> @simon-ess commented on this pull request.
> 
> Also, one other oddity I noticed: if you have the following as a db file:
> 
> record(ai, "FOO") {
>     field(INP, {pva: {pv: "BAR", proc: "CPP"}})
> }
> 
> record(aai, "BAR") {
>     field(NELM, 10)
> }
> then running pvput BAR "[1,2,3]" with an IOC running softIocPVA the linked ai record fetches the first value. However, with softIocPVX it does not.
> 
> In ioc/pvalink_link.cpp <https://github.com/mdavidsaver/pvxs/pull/57#discussion_r1383373488>:
> 
> > +    log_debug_printf(_logger, "%s type change V=%c S=%c N=%c S=%c M=%c\n",
> +                     plink->precord->name,
> +                     fld_value ? 'Y' : 'N',
> +                     fld_seconds ? 'Y' : 'N',
> +                     fld_nanoseconds ? 'Y' : 'N',
> +                     fld_severity ? 'Y' : 'N',
> +                     fld_meta ? 'Y' : 'N');
> I'm a bit confused by this debug statement: isn't root["value"] the actual value? This debug message seems to be trying to say that the value/timestamp/etc. are being modified, but that's not how I read the code...
> 
> In ioc/pvalink_channel.cpp <https://github.com/mdavidsaver/pvxs/pull/57#discussion_r1383563635>:
> 
> > +            atomic_lock = ioc::DBManyLock(atomicrecs);
> +            atomic_records = std::move(atomic);
> +            nonatomic_records = std::move(nonatomic);
> +
> +            links_changed = false;
> +        }
> +
> +        update_seq++;
> +        update_evt.signal();
> +        log_debug_printf(_logger, "%s Sequence point %u\n", key.first.c_str(), update_seq);
> +    }
> +    // unlock link
> +
> +    if(!atomic_records.empty()) {
> +        ioc::DBManyLocker L(atomic_lock);
> This is perhaps an odd pedantic point, and probably out of scope... but aren't these names semantically backwards? It seems to me that DBManyLock should be doing the action, while DBManyLocker is the thing that does the action.
> 
> Not that this affects the code, but it gives me a bit of a hiccup on reading, as it looks like the lock is taken on line 416 (and not obviously ever let go), but it is in fact taken on line 430 (and let go shortly thereafter).
> 
> Looking at the Git Blame, it seems that we can blame @george-mcintyre <https://github.com/george-mcintyre>...
> 

Yeah, like employee and employer,  DBManyLock is the actual subject/object - the actual lock object, and DBManyLocker is the think that does the locking.  Though I can see how it looks confusing.

File: ioc/pvalink_link.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        fld_meta = std::move(root);
+    }
+
+    log_debug_printf(_logger, "%s type change V=%c S=%c N=%c S=%c M=%c\n",
+                     plink->precord->name,
+                     fld_value ? 'Y' : 'N',
+                     fld_seconds ? 'Y' : 'N',
+                     fld_nanoseconds ? 'Y' : 'N',
+                     fld_severity ? 'Y' : 'N',
+                     fld_meta ? 'Y' : 'N');
```
[the remark this answers is not in the document]
Comment (mdavidsaver) [reply ioc/pvalink_link.cpp:118]: This code, in `pvaLink::onTypeChange()`, runs when a link (re)connects and the structure of the initial Value is being inspected.  This debug print gives a summary of the results of this inspection.  Nothing is being modified.

wrt. `root["value"]` vs. `fld_value`.  Consider a hypothetical case where link which targets `field: "timeStamp.userTag"`.  In this case `fld_value` will point to the userTag, and the other `fld_*` will be empty.

**Comment by mdavidsaver** [comment]
> Also, one other oddity I noticed ...

You may have meant to set `FTVL` on the `aai` record, which despite the name defaults to `STRING` (imo. a better name would be `arrin`).  So this case is trying to read `string[]` into `DBR_DOUBLE`, which is a TODO.  Currently `string[]` can only be read into `DBR_STRING`.

**Comment by mdavidsaver** [comment]
Another update.  Adds `string[]` conversions on GET.  Also a significant reorganization of the initHooks and atexit sequencing.  Now the four sequences (server, single pv, group pv, and pvalink) are combined together into one.  I have also extended the dbUnitTest.h helper APIs in `pvxs/iochooks.h` to (hopefully) allow for external testing w/ QSRV.

At this point, I have no more major TODOs.

**Comment by mdavidsaver** [comment]
I have found an issue where input links with CPP don't get an initial scan if the associated Channel is already connected and idle.  This is annoyingly similar to a previous issue with onTypeChange() (https://github.com/mdavidsaver/pvxs/pull/57#issuecomment-1756704551).

File: ioc/pvalink_link.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (simon-ess) [reply ioc/pvalink_link.cpp:119]: My question is more about the line `fld_value ? "Y" : "N"`. If `fld_value` is the actual value, isn't this going to only produce `Y` if the value is non-zero?

**Comment by simon-ess** [comment]
@george-mcintyre I think that part of my confusion comes from the code in PVXS using the `Guard` object, but now that I think of it more in that case, `Guard` is both noun and verb, so maybe I've just been reading it backwards from everyone else the whole time... 🙃

**Comment by simon-ess** [comment]
> > Also, one other oddity I noticed ...
> 
> You may have meant to set `FTVL` on the `aai` record, which despite the name defaults to `STRING` (imo. a better name would be `arrin`). So this case is trying to read `string[]` into `DBR_DOUBLE`, which is a TODO. Currently `string[]` can only be read into `DBR_STRING`.

I think my question is more that I would have expected softIovPVA and softIocPVX to behave the same with the same .db file, but they do not.

File: ioc/pvalink_link.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdavidsaver) [reply ioc/pvalink_link.cpp:119]: `fld_value` is meant to behave like a pointer to an actual value.  cf. `Value::operator bool()`.

**Comment by mdavidsaver** [comment]
> ... I would have expected softIovPVA and softIocPVX to behave the same ...

This is the expectation I would like you to have.  I was only making a point about why the were different.  (this difference should be fixed)  Maybe I am the only one who keeps forgetting that `aai` defaults to `STRING`...

File: ioc/pvalink_link.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (simon-ess) [reply ioc/pvalink_link.cpp:119]: Ah, that's right. Thanks for the clarification.

**Comment by mdavidsaver** [comment]
> subtitles and nuisances of naming

Ah, of course I meant nuances...

**Comment by mdavidsaver** [comment]
For some reason `testqsingle` is now failing on all appveyor jobs, but not on the GHA windows jobs.

```
testqsingle.tap .. 
not ok 89 - Instance leak ServerPvt : 1
not ok 90 - Instance leak ServerSource : 1
not ok 91 - Instance leak StructTop : 1
not ok 92 - Instance leak UDPListener : 1
not ok 93 - Instance leak evbase : 2
not ok 94 - Instance leak evbaseRunning : 2
All 88 subtests passed 
```

These counts point to a `Server` being orphaned somehow.  The trigger is likely my changing testqsingle to stop and restart the IOC.  Why this only manifests on appveyor is at this point a mystery to me.

**Comment by mdavidsaver** [comment]
The (re)initialization issue which was causing leaks and (sometimes) test failures should now be fixed.

**Comment by mdavidsaver** [comment]
Ok.  I think this is good enough.  Thanks @simon-ess .

**Pull request merged**